### PR TITLE
gh-104050: Annotate Argument Clinic return converters

### PR DESCRIPTION
**Pull request by erlend-aasland**
* Issue: gh-104050

### CODE REVIEW

File: Tools/clinic/clinic.py [hunk 1]
```diff
@@ -3005,6 +2996,14 @@ def parser_name(self):
 # note however that they will never be called with keyword-only parameters.
 legacy_converters: ConverterDict = {}
 
+# maps strings to callables.
+# these callables must be of the form:
+#   def foo(*, ...)
+# The callable may have any number of keyword-only parameters.
+# The callable must return a CConverter object.
+# The callable should not call builtins.print.
+return_converters: ConverterDict = {}
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py]: This doesn't look right to me. The `converters` dict above is a mapping of `{str: Callable[..., CConverter}`. But this dictionary is a mapping of `{str: Callable[..., CReturnConverter}`. `CReturnConverter` is a different class to `CConverter`.

File: Tools/clinic/clinic.py [hunk 2]
```diff
@@ -3966,8 +3965,10 @@ def set_template_dict(self, template_dict):
             template_dict['base_type_ptr'] = type_ptr
 
 
-
-def add_c_return_converter(f, name=None):
+def add_c_return_converter(
+        f: type,
+        name: str | None = None
+) -> ConverterType:
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py]: This is incorrect. `ConverterType` is an alias for `Callable[..., CConverterType]`. But the `return_converters` dictionary has subclasses of `CCReturnConverter` as its values, which is a different class.

**Review by AlexWaygood — CHANGES_REQUESTED** [review]

File: Tools/clinic/clinic.py (hunk 1 above)
Comment (erlend-aasland) [reply Tools/clinic/clinic.py]: Good catch. I'll fix later.

File: Tools/clinic/clinic.py [hunk 3]
```diff
@@ -3978,9 +3981,15 @@ def add_c_return_converter(f, name=None):
 
 
 class CReturnConverterAutoRegister(type):
-    def __init__(cls, name, bases, classdict):
+    def __init__(
+            cls: ReturnConverterType,
+            name: str,
+            bases: tuple[type, ...],
+            classdict: dict[Any, Any]
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py]: It's true that, strictly speaking, it's _possible_ to pass a dictionary with non-`str` keys in here, but it's almost certainly not what you want ;)

```pycon
>>> t = type('t', (), {'foo': 'bar', 3: 3})
>>> t
<class '__main__.t'>
>>> t.foo
'bar'
>>> t.3
  File "<stdin>", line 1
    t.3
     ^^
SyntaxError: invalid syntax
```

```suggestion
            classdict: dict[str, Any]
```

**Review by AlexWaygood — COMMENTED** [review]
Looks good other than one other small point -- feel free to merge once that's resolved!

File: Tools/clinic/clinic.py (hunk 3 above)
Comment (erlend-aasland) [reply Tools/clinic/clinic.py]: Thanks! I'm not sure why I put `Any, Any` there; oh, well.

**Review by AlexWaygood — APPROVED** [review]

**Pull request merged**